src/game/client/ingameui.cpp

```diff
@@ -63,3 +1158,597 @@ bool InGameUI::Are_Selected_Objects_Controllable()
+    Object *obj = nullptr;
+
+    if (build_drawable != nullptr) {
+        obj = build_drawable->Get_Object();
+    }
+
+    if (obj != nullptr) {
+        m_pendingPlaceSourceObjectID = obj->Get_ID();
+    }
+
+    if (g_theMouse != nullptr) {
+        if (build != nullptr) {
+            m_mouseMode = MOUSEMODE_BUILD_PLACE;
+            m_mouseCursor = CURSOR_TARGET;
+            g_theMouse->Capture();
+            Set_Mouse_Cursor(CURSOR_TARGET);
+            Drawable *drawable = g_theThingFactory->New_Drawable(build, DRAWABLE_STATUS_8);
+
+            if (obj != nullptr) {
+                int color;
+
+                if (g_theWriteableGlobalData->m_timeOfDay == TIME_OF_DAY_NIGHT) {
+                    color = obj->Get_Controlling_Player()->Get_Night_Color();
+                } else {
+                    color = obj->Get_Controlling_Player()->Get_Color();
+                }
+
+                drawable->Set_Indicator_Color(color);
+            }
+
+            captainslog_dbgassert(
+                drawable != nullptr, "Unable to create icon at cursor for placement '%s'", build->Get_Name().Str());
+            drawable->Set_Orientation(build->Get_Placement_View_Angle());
+            drawable->Set_Opacity(0.44999999f);
+            captainslog_dbgassert(*m_placeIcon == nullptr, "Place_Build_Available, build icon array is not empty!");
+            *m_placeIcon = drawable;
+        } else {
+            if (m_mouseMode == MOUSEMODE_BUILD_PLACE) {
+                m_mouseMode = MOUSEMODE_DEFAULT;
+                m_mouseCursor = CURSOR_ARROW;
+            }
+
+            g_theMouse->Release_Capture();
+            Set_Mouse_Cursor(CURSOR_ARROW);
+            Set_Placement_Start(nullptr);
+            Destroy_Placement_Icons();
+
+            if (obj != nullptr) {
+                ProductionUpdateInterface *production = obj->Get_Production_Update_Interface();
+
+                if (production != nullptr) {
+                    production->Set_Special_Power_Construction_Command_Button(nullptr);
+                }
+            }
+        }
+    }
+}
+
+void InGameUI::Destroy_Placement_Icons()
+{
+    for (int i = 0; i < g_theWriteableGlobalData->m_maxLineBuildObjects; i++) {
+        if (m_placeIcon[i] != nullptr) {
+            g_theTerrainVisual->Remove_Faction_Bib_Drawable(m_placeIcon[i]);
+            g_theGameClient->Destroy_Drawable(m_placeIcon[i]);
+            m_placeIcon[i] = nullptr;
+        }
+    }
+
+    g_theTerrainVisual->Remove_All_Bibs();
+}
+
+void InGameUI::Recreate_Control_Bar()
+{
+    GameWindow *window =
+        g_theWindowManager->Win_Get_Window_From_Id(nullptr, g_theNameKeyGenerator->Name_To_Key("ControlBar.wnd"));
+
+    if (window != nullptr) {
+        window->Delete_Instance();
+    }
+
+    m_idleWorkerWin = nullptr;
+    Create_Control_Bar();
+
+    if (g_theControlBar != nullptr) {
+        delete g_theControlBar;
+        g_theControlBar = new ControlBar();

```

Comment:
Not outside scope? This means it will only create `ControlBar` if it existed before.

Reply:
This is correct.